src/Utility/Security.php

```diff
@@ -121,15 +129,42 @@ public static function randomBytes(int $length): string
      * Creates a secure random string.
      *
      * @param int $length String length. Default 64.
+     * @param string $includes character sets to be included in the random string,
+     *                          each set will be represented by at least one character
      * @return string
      */
-    public static function randomString(int $length = 64): string
+    public static function randomString(int $length = 64, string ...$includes): string
     {
-        return substr(
-            bin2hex(Security::randomBytes((int)ceil($length / 2))),
-            0,
-            $length
-        );
+        $string = '';
+
+        if ($length === 0) {
+            return $string;
+        }
+
+        if ($length < count($includes)) {
+            throw new InvalidArgumentException('Length must not be lower than number of character sets.');
+        }
+
+        $remaining = count($includes);
+        foreach ($includes as $set) {
+            $remaining--;
+
+            $characters = random_int(1, ($length - strlen($string) - $remaining));
+
+            for ($i=0; $i<$characters; $i++) {
+                $string .= $set[random_int(0, strlen($set) - 1)];
+            }
+        }
+
+        if ($length > strlen($string)) {
+            $string .= substr(
+                bin2hex(Security::randomBytes((int)ceil(($length - strlen($string)) / 2))),
+                0,
+                ($length - strlen($string))
+            );
+        }
+
+        return str_shuffle($string);
```

Comment:
We can't do this. Remember `randomString()` is for security sensitive scenarios. The documentation for php mentions this is not a scenario where `str_shuffle` should be used.

> This function does not generate cryptographically secure values, and must not be used for cryptographic purposes, or purposes that require returned values to be unguessable. 
> https://www.php.net/str_shuffle

Instead of modifying the existing `randomString`, we should add a new method. That will let us add parameters to control the minimum number of characters per class so that more scenarios could be covered.

Reply:
> We can't do this. Remember `randomString()` is for security sensitive scenarios. The documentation for php mentions this is not a scenario where `str_shuffle` should be used.
> 
> > This function does not generate cryptographically secure values, and must not be used for cryptographic purposes, or purposes that require returned values to be unguessable.
> > https://www.php.net/str_shuffle

I could argue that since the string to be shuffled is composed in cryptographically secure way, shuffling doesn't decrease its security.

But if you would rather have it separate, I have no issue with making it `randomPassword` or something like that.

Reply:
> I could argue that since the string to be shuffled is composed in cryptographically secure way, shuffling doesn't decrease its security.

Yes, it does - potentially.

Remember a chain is only as strong as its weakest link.

And since the non-cryptographically secure shuffling is part of the chain to the resulting string, the string itself must be viewed as such.

It might not matter that much for a single password but it certainly does for a series of generated passwords.

Plus, in security one does not take chances.

Reply:
> Remember a chain is only as strong as its weakest link.

Not relevant in this case. Zero predictability of the cryptographically secure randomness acts like multiplication by zero. 

You can have so many predictable parts, but when you add randomness, it's all random. Even when you do anything predictable with random, it's still random. Like multiply by zero, it's still zero.

Anyway, my previous comment is still valid. I am fine with `randomString()` staying as-is (only lowercase and numbers) and I still see value in having more complex random string generator, that can have different name.